cli/listener/src/main.rs

```diff
@@ -210,74 +222,396 @@ pub fn get_apy_for_period(
+            // Update metrics snapshot.
+            *self.metrics_snapshot.lock().unwrap() = Arc::new(self.metrics.clone());
+            std::thread::sleep(sleep_time);
+        }
+    }
+
+    fn get_sleep_time_after_error(&mut self) -> Duration {
+        // For the sleep time we use exponential backoff with jitter [1]. By taking
+        // the time since the last success as the target sleep time, we get
+        // exponential backoff. We clamp this to ensure we don't wait indefinitely.
+        // 1: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+        let time_since_last_success = self.last_read_success.elapsed();
+        let min_sleep_time = Duration::from_secs_f32(0.2);
+        let max_sleep_time = Duration::from_secs_f32(300.0);
+        let target_sleep_time = time_since_last_success.clamp(min_sleep_time, max_sleep_time);
+        let sleep_time = self
+            .rng
+            .gen_range(Duration::from_secs(0)..target_sleep_time);
+        println!("Sleeping {:?} after error ...", sleep_time);
+        sleep_time
+    }
+
+    pub fn get_sleep_time(&self) -> Duration {
+        // Sleep until is time to get the next exchange rate.
+        let sleep_time = std::time::Duration::from_secs(self.opts.poll_interval_seconds as u64);
+        println!(
+            "Sleeping for {:?} after getting the Solido exchange rate",
+            sleep_time
+        );
+        sleep_time
+    }
+}
+
+#[derive(Clone)]
+struct Metrics {
+    /// Number of times that we checked the price.
+    polls: u64,
+
+    /// Number of times that we tried to get the exchange rate, but encountered an error.
+    errors: u64,
+
+    /// Solido's maximum price interval.
+    solido_average_30d_interval_price: Option<IntervalPrices>,
+}
+
+impl Metrics {
+    pub fn write_prometheus<W: std::io::Write>(&self, out: &mut W) -> std::io::Result<()> {
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solido_pricedb_polls_total",
+                help: "Number of times we polled the exchange rate, since launch.",
+                type_: "counter",
+                metrics: vec![Metric::new(self.polls)],
+            },
+        )?;
+        write_metric(out, &MetricFamily {
+            name: "solido_pricedb_errors_total",
+            help: "Number of times we encountered an error while trying to get the exchange rate, since launch.",
+            type_: "counter",
+            metrics: vec![Metric::new(self.errors)]
+        })?;
+        if let Some(interval_price) = &self.solido_average_30d_interval_price {
+            write_metric(
+                out,
+                &MetricFamily {
+                    name: "solido_pricedb_30d_average_apy",
+                    help: "Average 30d APY",
+                    type_: "gauge",
+                    metrics: vec![Metric::new(interval_price.annual_percentage_rate())
+                        .with_label("solido", "APY".to_string())],
```

Comment:
What is the point of this label? Labels are key-value pairs that you attach to a metric to distinguish multiple instances of the same metrics. For example, you might have two instances of an http server, each with their own request counter, and you can have labels `instance="server1"` and `instance="server2"`.

In this particular case, if in the future we want to support multiple pools, it would make sense to add a `pool="solido"` label. Then you can query `solido_pricedb_30d_average_apy` in Prometheus or Grafana, and it will return the 30d APYs of all pools as separate series in the same chart.

Reply:
Yes... For now we can use it without a label, since we have only Solido.